gettsim/parameter_vizualisation_simon/bokeh_app/Scripts/tax_rate_interact.py

```diff
@@ -0,0 +1,127 @@
+import numpy as np
+import pandas as pd
+from bokeh.layouts import column
+from bokeh.models import ColumnDataSource
+from bokeh.models import NumeralTickFormatter
+from bokeh.models import Panel
+from bokeh.models import Slider
+from bokeh.plotting import figure
+
+from gettsim import set_up_policy_environment
+from gettsim.piecewise_functions import piecewise_polynomial
+from gettsim.taxes.eink_st import st_tarif
+
+
+def tax_rate_interact(plot_dict):
+    plot_dict = plot_dict["tax_rate"]
+
+    def prepare_data(start, end):
+        """
+        For a given year span returns the policy parameters to plot income tax
+        rate per income
+
+        sel_year (Int): The year for which the data will be simulated. The range for
+                        which parameters can be simulated is 2002-2020.
+
+        returns dict
+        """
+        years = range(start, end)
+        einkommen = pd.Series(data=np.linspace(0, 300000, 250))
+        tax_rate_dict_full = {}
+        for i in years:
+            policy_params, policy_functions = set_up_policy_environment(i)
+            eink_params = policy_params["eink_st"]
+            soli_params = policy_params["soli_st"]["soli_st"]
+
+            eink_tax = st_tarif(einkommen, eink_params)
+            soli = piecewise_polynomial(
+                eink_tax,
+                thresholds=soli_params["thresholds"],
+                rates=soli_params["rates"],
+                intercepts_at_lower_thresholds=soli_params[
+                    "intercepts_at_lower_thresholds"
+                ],
+            )
+            marginal_rate = np.gradient(eink_tax, einkommen)
+
+            tax_rate_dict_full[i] = {
+                "tax_rate": (eink_tax / einkommen),
+                "overall_tax_rate": ((soli + eink_tax) / einkommen),
+                "marginal_rate": pd.Series(marginal_rate),
+                "income": einkommen,
+            }
+
+        return tax_rate_dict_full
+
+    def make_dataset(sel_year, tax_rate_dict_full):
+        dataset = tax_rate_dict_full[sel_year]
+
+        return ColumnDataSource(dataset)
+
+    def plotstyle(p, plot_dict):
```

Comment:
Ideally, you would of course only need one function `plotstyle` which is used for every graph.

Reply:
Yes I agree. I will look into it and push it with the next update..

Reply:
I moved the plot styling into a separate module, which can now be used for all plots. I'm still looking into the data preparation stuff but for now the start-up time of the server app is not ideal because it needs to recreate the data sources everytime when initiated.

Reply:
Maybe I am misunderstanding something about the architecture. But I thought the server app would just run forever, so start-up time is a one-time cost? Or is a new instance created each time a user plays around with a plot?

Reply:
>  I'm still looking into the data preparation stuff but for now the start-up time of the server app is not ideal because it needs to recreate the data sources everytime when initiated.

But isn't it still cheaper to load all the data everytime from `.csv` rather than loop through all the `.yaml` files?

Reply:
> 
> 
> Maybe I am misunderstanding something about the architecture. But I thought the server app would just run forever, so start-up time is a one-time cost? Or is a new instance created each time a user plays around with a plot?

At least locally if I don't explicitly close the server it runs forever and I can just go back and play with the plot.. However if I understand the bokeh user guide correctly it might be different if the server is hosted:
> The application code is executed in the Bokeh server every time a new connection is made, to create the new Bokeh Document that will be synced to the browser. 
https://docs.bokeh.org/en/latest/docs/user_guide/server.html

Reply:
Fine. Just don't worry about it. Remember [Knuth](http://wiki.c2.com/?PrematureOptimization).

That is, as long as we have a load_data() function upfront, we can always change its behaviour. E.g., create a version with all parameters and pickle it when firing up the server / updating GETTSIM. When a connection is made, just load the pickle. That sort of refactoring are near-trivial changes if the tasks are well separated (load all data first, then process it, then plot).

Reply:
Okay cool, we have a clear functional separation between data preparation, selecting from the prepared data, plotting and then updating the plot based on sliders.